File: gto-support-api-base/src/main/java/org/ccci/gto/android/common/api/TheKeySession.java
package org.ccci.gto.android.common.api;

import android.content.SharedPreferences;
import android.text.TextUtils;

import java.util.Arrays;
import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class TheKeySession extends Session {
    @Nullable
    private final String mGuid;

    public TheKeySession(@Nullable final String id, @Nullable final String guid) {
        this(id, guid, PREF_SESSION_BASE_NAME);
    }

    public TheKeySession(@Nullable final String id, @Nullable final String guid, @NonNull final String baseAttrName) {
        super(id, (guid != null ? guid.toUpperCase(Locale.US) + "." : "") + baseAttrName);
        mGuid = guid != null ? guid.toUpperCase(Locale.US) : null;
    }

    public TheKeySession(@NonNull final SharedPreferences prefs, @Nullable final String guid) {
        this(prefs, guid, PREF_SESSION_BASE_NAME);
    }

    public TheKeySession(@NonNull final SharedPreferences prefs, @Nullable final String guid,
                         @NonNull final String baseAttrName) {
        super(prefs, (guid != null ? guid.toUpperCase(Locale.US) + "." : "") + baseAttrName);
        mGuid = guid != null ? guid.toUpperCase(Locale.US) : null;
    }

    @Override
    public boolean isValid() {
        return super.isValid() && mGuid != null;
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        final TheKeySession that = (TheKeySession) o;
        return super.equals(o) && TextUtils.equals(mGuid, that.mGuid);
    }

    @Override
    public int hashCode() {
        return super.hashCode() * 31 + Arrays.hashCode(new Object[] {mGuid});
    }
}

File: gto-support-api-base/src/main/kotlin/org/ccci/gto/android/common/api/TheKeySession.kt
package org.ccci.gto.android.common.api

import android.content.SharedPreferences
import java.util.Locale

@Deprecated("Since v4.0.0, We should no longer be using The Key for authentication")
class TheKeySession private constructor(
    prefs: SharedPreferences?,
    id: String?,
    guid: String?,
    baseAttrName: String,
) : Session(prefs = prefs, id = id, baseAttrName = guid.sanitizeGuid(".") + baseAttrName) {
    private val guid = guid?.sanitizeGuid()

    @JvmOverloads
    constructor(
        id: String?,
        guid: String?,
        baseAttrName: String = PREF_SESSION_BASE_NAME,
    ) : this(prefs = null, id = id, guid = guid, baseAttrName = baseAttrName)

    @JvmOverloads
    constructor(
        prefs: SharedPreferences,
        guid: String?,
        baseAttrName: String = PREF_SESSION_BASE_NAME,
    ) : this(prefs = prefs, id = null, guid = guid, baseAttrName = baseAttrName)

    override val isValid get() = super.isValid && guid != null

    override fun equals(other: Any?) = when {
        this === other -> true
        !(other is TheKeySession && javaClass == other.javaClass) -> false
        else -> super.equals(other) && guid == other.guid
    }

    override fun hashCode() = super.hashCode() * 31 + (guid?.hashCode() ?: 0)
}

private fun String?.sanitizeGuid(suffix: String = "") = if (this != null) "${uppercase(Locale.US)}$suffix" else ""
